### PR TITLE
Fix synthesizeReceipt

### PR DESCRIPTION
**Pull request by gsouquet**
Fixes https://github.com/vector-im/element-web/issues/23827
Fixes https://github.com/vector-im/element-web/issues/23754
Fixes https://github.com/vector-im/element-web/issues/23847

This is still missing a couple of tests to ensure synthetic receipts are added when paginating through a thread timeline set. But otherwise should be ready for review

## Checklist

* [ ] Tests written for new code (and old code if feasible)
* [x] Linter and other CI checks pass
* [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Fix synthesizeReceipt ([\#2916](https://github.com/matrix-org/matrix-js-sdk/pull/2916)). Fixes vector-im/element-web#23827 vector-im/element-web#23754 and vector-im/element-web#23847.

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -9325,6 +9331,19 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
     }
 }
 
+export function getCount({ type, room, event }: {
```
Comment (t3chguy) [inline src/client.ts:9334]: Could this be moved to somewhere more logical, `getCount` in `client.ts` implies it gets a count of clients/a client, which it doesn't

File: src/models/thread.ts [hunk 2]
```diff
@@ -192,6 +192,18 @@ export class Thread extends ReadReceipt<EmittedEvents, EventHandlerMap> {
...
     };
 
+    private onTimelineEvent = (
+        event: MatrixEvent,
+        room: Room | undefined,
+        toStartOfTimeline: boolean | undefined,
+    ): void => {
+        // Add a synthesized receipt to all live events added to the timeline
+        if (!toStartOfTimeline) {
+            room!.addLocalEchoReceipt(event.getSender()!, event, ReceiptType.Read);
```
Comment (t3chguy) [inline src/models/thread.ts:202]: Won't this possibly clobber a main room timeline RR given that `threadId` prefers threads and falls back to `MAIN_ROOM_TIMLINE`

File: src/client.ts [hunk 3]
```diff
@@ -9325,6 +9331,19 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
...
+    type: NotificationCountType; room: Room; event: MatrixEvent;
+}): number {
+    const isThreadEvent = !!event.threadRootId && !event.isThreadRoot;
+
+    return (isThreadEvent
+        ? room.getThreadUnreadNotificationCount(
+            event.threadRootId,
+            type,
+        )
```
Comment (t3chguy) [inline src/client.ts:9343]:
```suggestion
        ? room.getThreadUnreadNotificationCount(event.threadRootId, type)
```

File: src/models/thread.ts (hunk 2 above)
Comment (gsouquet) [reply src/models/thread.ts:202]: I'm not sure I follow the scenario you're talking about, nor what the risk would be?

Comment (t3chguy) [reply src/models/thread.ts:202]: that a RR the thread root in the main timeline will vanish as soon as it's sender sends a message into the thread and a synthetic RR is generated with a `thread_id` which isn't `MAIN_ROOM_TIMELINE`

File: src/models/thread.ts [hunk 4]
```diff
@@ -192,6 +192,18 @@ export class Thread extends ReadReceipt<EmittedEvents, EventHandlerMap> {
         }
     };
 
+    private onTimelineEvent = (
+        event: MatrixEvent,
+        room: Room | undefined,
+        toStartOfTimeline: boolean | undefined,
+    ): void => {
+        // Add a synthesized receipt to all live events added to the timeline
+        if (!toStartOfTimeline) {
```
Comment (t3chguy) [inline src/models/thread.ts:201]: I don't think this means it is live, forward paginating will be included too, while that is fine I think this comment is not

**Review by t3chguy — APPROVED** [review]
:shipit:

**Pull request merged**